modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
+            // separate cache/clear API.
+            this.caches.put(diskCache, new TierInfo(isDiskCacheEnabled, TIER_DIMENSION_VALUE_DISK));
+            this.statsHolder.setDiskCacheEnabled(isDiskCacheEnabled);
+        }
+
+        @Override
+        public V get(ICacheKey<K> key) {
+            Tuple<V, String> cacheValueTuple = getValueFromTieredCache(true, false).apply(key);
             if (cacheValueTuple == null) {
-                // Only one of the threads will succeed putting a future into map for the same key.
-                // Rest will fetch existing future and wait on that to complete.
-                future = completableFutureMap.putIfAbsent(key, new CompletableFuture<>());
+                return null;
             }
+            return cacheValueTuple.v1();
         }
-        List<String> heapDimensionValues = statsHolder.getDimensionsWithTierValue(key.dimensions, TIER_DIMENSION_VALUE_ON_HEAP);
-        List<String> diskDimensionValues = statsHolder.getDimensionsWithTierValue(key.dimensions, TIER_DIMENSION_VALUE_DISK);
-
-        if (cacheValueTuple == null) {
-            // Add the value to the onHeap cache. We are calling computeIfAbsent which does another get inside.
-            // This is needed as there can be many requests for the same key at the same time and we only want to load
-            // the value once.
-            V value = compute(key, loader, future);
-            // Handle stats
-            if (loader.isLoaded()) {
-                // The value was just computed and added to the cache by this thread. Register a miss for the heap cache, and the disk cache
-                // if present
+
+        @Override
+        public void put(ICacheKey<K> key, V value) {
+            // First check in case the key is already present in either of tiers.
+            Tuple<V, String> cacheValueTuple = getValueFromTieredCache(true, false).apply(key);
+            if (cacheValueTuple == null) {
+                // In case it is not present in any tier, put it inside onHeap cache by default.
+                try (ReleasableLock ignore = writeLock.acquire()) {
+                    onHeapCache.put(key, value);
+                }
                 updateStatsOnPut(TIER_DIMENSION_VALUE_ON_HEAP, key, value);
-                statsHolder.incrementMisses(heapDimensionValues);
-                if (caches.get(diskCache).isEnabled()) {
-                    statsHolder.incrementMisses(diskDimensionValues);
+            } else {
+                // Put it inside desired tier.
+                try (ReleasableLock ignore = writeLock.acquire()) {
+                    for (Map.Entry<ICache<K, V>, TierInfo> entry : this.caches.entrySet()) {
+                        if (cacheValueTuple.v2().equals(entry.getValue().tierName)) {
+                            entry.getKey().put(key, value);
+                        }
+                    }
+                    updateStatsOnPut(cacheValueTuple.v2(), key, value);
+                }
```

Comment:
The behavior for put seems to be different than before. SegmentedCache should not be the reason for that I guess.

Reply:
Yeah this is unrelated to segmented cache. Earlier` put()` logic was incorrect, as it tries to put the value without checking whether its present in the cache and replacing it accordingly. So did club this change considering it was not substantial